### DISCUSSION TITLE
How to handle "Abnormal Termination"?

### DISCUSSION THREAD

**Discussion by qiankunli** [openai/whisper#2478] [Q&A]
What are the general causes for a program to throw an "Abnormal Termination" exception, such as insufficient memory?

Operating environment: Debian CPU Kubernetes pod (request=8g, limit=16g)

**Comment by Advait251206** [comment]
"Abnormal Termination" is a fairly generic error and usually means the process exited unexpectedly without Whisper itself raising a normal Python exception. In a Kubernetes environment, the root cause is often outside of Whisper.

## Common Causes

### 1. Out of Memory (OOM) — Most Common

Even if your pod has:

```yaml
requests:
  memory: 8Gi

limits:
  memory: 16Gi
```

the process can still be killed if:

- The model plus audio processing exceeds the limit.
- Multiple transcriptions run concurrently.
- Large audio files create large intermediate tensors.
- Other processes in the container consume memory.

Check:

```bash
kubectl describe pod <pod-name>
```

Look for:

```text
Reason: OOMKilled
Exit Code: 137
```

or:

```text
Container killed due to memory limit exceeded
```

### 2. CPU Resource Exhaustion

Whisper can be CPU-intensive, especially on long audio.

Symptoms:

- Process appears to hang.
- Kubernetes liveness/readiness probes fail.
- Container gets restarted.

Check:

```bash
kubectl top pod <pod-name>
```

and monitor CPU throttling.

---

### 3. Pod Eviction

The node itself may be under memory pressure.

Check:

```bash
kubectl describe pod <pod-name>
```

for messages like:

```text
Evicted
```

or:

```text
The node was low on resource: memory
```

---

### 4. FFmpeg Crash

Whisper relies on FFmpeg for audio decoding.

Possible issues:

- Corrupted audio file
- Unsupported codec
- Damaged container image
- FFmpeg process crash

Test separately:

```bash
ffmpeg -i audio.mp3 -f null -
```

If FFmpeg fails, Whisper may terminate before transcription starts.

---

### 5. Native Library Crash

Occasionally:

- PyTorch
- MKL/OpenMP
- BLAS libraries
- AVX instructions

can trigger a segmentation fault.

Typical exit codes:

```text
139
```

or

```text
Segmentation fault
```

Check container logs:

```bash
kubectl logs <pod-name> --previous
```

---

### 6. Large Audio Files

Whisper processes audio in chunks internally, but extremely large inputs can still create problems:

- Memory spikes
- Long-running jobs
- Timeouts

Consider pre-splitting very long audio files.

---

### 7. Kubernetes Timeouts

If transcription takes too long:

- Liveness probe may fail
- Readiness probe may fail
- Sidecars or ingress may timeout

Check:

```bash
kubectl describe pod <pod-name>
```

for restart events.

---

## How to Diagnose

### Check Pod Status

```bash
kubectl get pod <pod-name>
kubectl describe pod <pod-name>
```

### Check Previous Logs

```bash
kubectl logs <pod-name> --previous
```

### Check Exit Code

Common exit codes:

| Exit Code | Meaning |
|------------|----------|
| 137 | OOMKilled (SIGKILL) |
| 139 | Segmentation Fault |
| 143 | SIGTERM |
| 1 | Application Error |
| 134 | Abort Signal |

---

## Whisper-Specific Memory Estimates (CPU)

Approximate RAM requirements:

| Model | RAM |
|---------|---------|
| tiny | <1 GB |
| base | 1-2 GB |
| small | 2-4 GB |
| medium | 5-8 GB |
| large-v2/v3 | 10-16+ GB |

With concurrent requests, actual usage can be significantly higher.

---

## Recommendations

1. Check whether the container was **OOMKilled**.
2. Collect the pod **exit code** and previous logs.
3. Monitor memory usage during transcription.
4. Limit concurrent transcriptions if multiple requests share the same model instance.
5. For CPU deployments, consider using:
   - `small`
   - `medium`
   - `large-v3-turbo`
   
   instead of full `large-v3` if memory pressure is an issue.
6. If processing long recordings, split them into smaller chunks before transcription.

If you can provide:
- the full error message,
- pod exit code,
- Whisper model being used,
- audio duration,
- and the relevant Kubernetes events,

it will be much easier to determine whether the termination is due to OOM, FFmpeg, a native crash, or Kubernetes itself.
